plugins/ingest/src/main/java/org/elasticsearch/plugin/ingest/transport/IngestActionFilter.java

```diff
@@ -137,4 +151,91 @@ public void failed(Exception e) {
     public int order() {
         return Integer.MAX_VALUE;
     }
+
+    final static class BulkRequestSource implements Iterator<ActionRequest> {
+
+        final BulkRequest bulkRequest;
+        final ActionListener<BulkResponse> actionListener;
+
+        final Set<Integer> failedSlots;
+        final List<BulkItemResponse> itemResponses;
+
+        int currentSlot = -1;
+
+        BulkRequestSource(BulkRequest bulkRequest, ActionListener<BulkResponse> actionListener) {
+            this.bulkRequest = bulkRequest;
+            this.actionListener = actionListener;
+            this.failedSlots = new HashSet<>();
+            this.itemResponses = new ArrayList<>(bulkRequest.requests().size());
+        }
+
+        @Override
+        public ActionRequest next() {
+            return bulkRequest.requests().get(++currentSlot);
+        }
+
+        @Override
+        public boolean hasNext() {
+            return (currentSlot + 1) < bulkRequest.requests().size();
+        }
+
+        Tuple<BulkRequest, ActionListener<BulkResponse>> bulkRequestAndActionListener() {
```

Comment:
I agree that the tuple itself isn't nice, but I'm not sure if moving this logic outside the `BulkRequestSource` makes things cleaner. The `BulkRequestSource` is also used to keep track of failed bulk items caused by pipeline errors. That information is then later used to alter the bulk request and wrap the action listener if needed. If we move this logic outside of `BulkRequestSource` then this class doesn't have added value and might be replaced by an iterator we build from the bulk request itself (`bulkRequest.requests().iterator()`)

What if `bulkRequestAndActionListener()` is replaced by two methods (`getBulkRequests()` and `getActionListener()`). I kind like to keep this `BulkRequestSource` where we fetch the bulk items from and process failures.